### ISSUE TITLE
Unstable training performance

### ISSUE THREAD

**Issue by ocesp98** [stellargraph/stellargraph#1310]
I am using Stellargraph 0.8.3 and Tensorflow 1.14 and Python 3.6. When training the model, I get a validation AUC of around 0.8. When my colleague runs the same code, he automatically gets an AUC of 0.9. We're using exactly the same code and the same tensorflow and stellargraph versions, only he always gets a significantly higher AUC than I am. 
I also noticed that the training time is not stable. Sometimes it takes double as much epochs to achieve a worse performance or a slightly better one whereas my colleague already gets good results after a few epochs (and training time stays stable).
We don't know why this happens given that we use almost the same package environment and given that the code is exactly the same.
Do you have any idea what the issue could be?
Thank you in advance!
Kind regards

**Comment by ocesp98** [comment]
So, he did not use the workers parameter, whereas I did. The performance then somehow increased which I think is quite strange. Can it be that there is some kind of problem with the workers parameter that affects the training of the model?

**Comment by huonw** [comment]
Hi, this does sound peculiar. Could you say a bit more about the model you're creating? What algorithm and generator are you using?

**Comment by ocesp98** [comment]
We actually used HinSAGE as comparison model, but we think the same problem also happens with GraphSAGE. For HinSAGE, we are using the HinSAGEnodegenerator.

**Comment by ocesp98** [comment]
I now ran it on GraphSAGE and here as well I got a better performance (on the test as well as validation set) when removing the workers parameter. For GraphSAGE, I use the GraphSAGEnodegenerator

**Comment by kjun9** [comment]
Hi @ocesp98

Based on your comments, I went through a few runs with the GraphSAGE Cora Node classification notebook (it also uses GraphSAGE with GraphSAGENodeGenerator) with both the latest version of stellargraph and 0.8.3, but I haven't been able to reproduce your problem unfortunately. So the problem so far seems specific to your setup - would you be able to provide some kind of minimal example to reproduce the issue?

> Can it be that there is some kind of problem with the workers parameter that affects the training of the model?

Could be! But theoretically, `workers` parameter should just be pulling in multiple batches from the sequence object in parallel without affecting how those batches are processed by the model. So it would indicate it's either a bug in the tensorflow version you're using, or a bug in the way the stellargraph sequence object is handling when `__getitem__` gets called in parallel.

So I might also suggest try upgrading stellargraph which uses tensorflow 2.1 if that's possible for you? On the stellargraph side, I can think of at least this issue #508 being resolved since 0.8.3 that might affect how our sequence objects handle the `workers` parameter.

**Comment by ocesp98** [comment]
Another colleague upgraded Stellargraph to version 0.9 and used Tensorflow 2.0, however, training time became much slower. 

This is the code I use to build the model.


```
stelgr = sg.StellarGraph(H, node_features = node_features)
batch_size = 128; num_samples = [10, 5] 
generator = GraphSAGENodeGenerator(stelgr, batch_size, num_samples)
train_gen = generator.flow(train_data.index, train_targets, shuffle=True)

from tensorflow.keras import initializers
graphsage_model = GraphSAGE(
    aggregator = MeanPoolingAggregator,
    layer_sizes=[32, 32],
    generator = train_gen,
    bias=True,
    dropout=0.4,
    kernel_initializer = initializers.he_normal(seed=None)
)

x_inp, x_out = graphsage_model.build()
prediction = layers.Dense(units=1, activation="sigmoid")(x_out)

from tensorflow.keras import callbacks
callbacks_list = [
    callbacks.EarlyStopping(
    monitor = 'val_f1',
    patience = 5,
    mode='max',
    restore_best_weights = True) 
]


from tensorflow.keras.metrics import binary_accuracy, AUC, Recall
from tensorflow.keras.losses import binary_crossentropy

model = Model(inputs=x_inp, outputs=prediction)
model.compile(
    optimizer= optimizers.Adam(lr=1e-3),
    loss= binary_crossentropy,
    metrics=['AUC', binary_accuracy, "Recall", f1]
)
history = model.fit_generator(
    train_gen,
    epochs=30,
    verbose=1,
   # use_multiprocessing=True,
    validation_data=val_gen,
    callbacks = callbacks_list,
    shuffle=False
)
```

**Comment by kjun9** [comment]
@ocesp98 Sorry for not responding sooner! Were you able to find a solution to the problem in the end? 

I tweaked your example a bit to run on the Cora dataset while trying to keep the gist of the code roughly the same, but unfortunately wasn't able to reproduce your problem. I'm suspecting it might be specific to your dataset, although it is strange that the `workers` parameter has anything to do with it 🤔 

<details>
  <summary>Here's what I tried, for reference</summary>
  
Some setup code
```python
import stellargraph as sg
from stellargraph import datasets
from stellargraph.mapper import GraphSAGENodeGenerator
from stellargraph.layer import GraphSAGE, MeanPoolingAggregator
import tensorflow as tf
from tensorflow.keras import layers
from tensorflow.keras import Model
from tensorflow import optimizers
from sklearn import preprocessing, feature_extraction, model_selection

cora = datasets.Cora()
stelgr, node_subjects = cora.load()

train_subjects, test_subjects = model_selection.train_test_split(
    node_subjects, train_size=0.7, test_size=None, stratify=node_subjects
)

target_encoding = preprocessing.LabelBinarizer()
train_targets = target_encoding.fit_transform(train_subjects)
test_targets = target_encoding.transform(test_subjects)
```

Adapted example (categorical instead of binary classification):
```python
batch_size = 128; num_samples = [10, 5] 
generator = GraphSAGENodeGenerator(stelgr, batch_size, num_samples)
train_gen = generator.flow(train_targets.index, train_targets, shuffle=True)

from tensorflow.keras import initializers
graphsage_model = GraphSAGE(
    aggregator = MeanPoolingAggregator,
    layer_sizes=[32, 32],
    generator = generator,
    bias=True,
    dropout=0.4,
    kernel_initializer = initializers.he_normal(seed=None)
)

x_inp, x_out = graphsage_model.in_out_tensors()
prediction = layers.Dense(units=train_targets.shape[1], activation="softmax")(x_out)

from tensorflow.keras import callbacks
callbacks_list = [
    callbacks.EarlyStopping(
    monitor = 'val_acc',
    patience = 5,
    mode='max',
    restore_best_weights = True)
]


from tensorflow.keras.metrics import binary_accuracy, AUC, Recall
from tensorflow.keras.losses import binary_crossentropy, categorical_crossentropy

model = Model(inputs=x_inp, outputs=prediction)
model.compile(
    optimizer= optimizers.Adam(lr=1e-3),
    loss= categorical_crossentropy,
    metrics=['AUC', "acc", "Recall"]
)

test_gen = generator.flow(test_subjects.index, test_targets)

history = model.fit(
    train_gen,
    epochs=30,
    verbose=1,
    validation_data=test_gen,
    callbacks = callbacks_list,
    # workers=4
    shuffle=False
)
```

Output using `workers=4`
```
Epoch 1/30
15/15 [==============================] - 3s 197ms/step - loss: 1.8607 - auc: 0.6451 - acc: 0.2770 - recall: 0.0000e+00 - val_loss: 1.7042 - val_auc: 0.7855 - val_acc: 0.3875 - val_recall: 0.0000e+00
Epoch 2/30
15/15 [==============================] - 3s 185ms/step - loss: 1.6962 - auc: 0.7978 - acc: 0.4285 - recall: 0.0000e+00 - val_loss: 1.5570 - val_auc: 0.8709 - val_acc: 0.5277 - val_recall: 0.0000e+00
Epoch 3/30
15/15 [==============================] - 3s 167ms/step - loss: 1.5742 - auc: 0.8749 - acc: 0.5720 - recall: 0.0000e+00 - val_loss: 1.4298 - val_auc: 0.9247 - val_acc: 0.6851 - val_recall: 0.0000e+00
Epoch 4/30
15/15 [==============================] - 3s 170ms/step - loss: 1.4573 - auc: 0.9207 - acc: 0.6707 - recall: 0.0000e+00 - val_loss: 1.3169 - val_auc: 0.9465 - val_acc: 0.7601 - val_recall: 0.0000e+00
Epoch 5/30
15/15 [==============================] - 3s 174ms/step - loss: 1.3462 - auc: 0.9517 - acc: 0.7483 - recall: 0.0000e+00 - val_loss: 1.2195 - val_auc: 0.9610 - val_acc: 0.8057 - val_recall: 0.0000e+00
Epoch 6/30
15/15 [==============================] - 3s 178ms/step - loss: 1.2401 - auc: 0.9675 - acc: 0.8074 - recall: 0.0000e+00 - val_loss: 1.1371 - val_auc: 0.9650 - val_acc: 0.8327 - val_recall: 0.0000e+00
Epoch 7/30
15/15 [==============================] - 3s 169ms/step - loss: 1.1618 - auc: 0.9727 - acc: 0.8396 - recall: 0.0000e+00 - val_loss: 1.0566 - val_auc: 0.9687 - val_acc: 0.8438 - val_recall: 0.0234
Epoch 8/30
15/15 [==============================] - 3s 168ms/step - loss: 1.0806 - auc: 0.9795 - acc: 0.8670 - recall: 0.0185 - val_loss: 0.9933 - val_auc: 0.9734 - val_acc: 0.8598 - val_recall: 0.1673
Epoch 9/30
15/15 [==============================] - 3s 170ms/step - loss: 1.0092 - auc: 0.9845 - acc: 0.8828 - recall: 0.1050 - val_loss: 0.9390 - val_auc: 0.9741 - val_acc: 0.8499 - val_recall: 0.3100
Epoch 10/30
15/15 [==============================] - 3s 183ms/step - loss: 0.9545 - auc: 0.9845 - acc: 0.8897 - recall: 0.2185 - val_loss: 0.8871 - val_auc: 0.9759 - val_acc: 0.8672 - val_recall: 0.4699
Epoch 11/30
15/15 [==============================] - 3s 181ms/step - loss: 0.8883 - auc: 0.9880 - acc: 0.9077 - recall: 0.3346 - val_loss: 0.8427 - val_auc: 0.9770 - val_acc: 0.8745 - val_recall: 0.5437
Epoch 12/30
15/15 [==============================] - 3s 181ms/step - loss: 0.8305 - auc: 0.9904 - acc: 0.9182 - recall: 0.4406 - val_loss: 0.7883 - val_auc: 0.9802 - val_acc: 0.8844 - val_recall: 0.6125
Epoch 13/30
15/15 [==============================] - 3s 175ms/step - loss: 0.7891 - auc: 0.9893 - acc: 0.9172 - recall: 0.5340 - val_loss: 0.7602 - val_auc: 0.9795 - val_acc: 0.8795 - val_recall: 0.6494
Epoch 14/30
15/15 [==============================] - 2s 165ms/step - loss: 0.7440 - auc: 0.9922 - acc: 0.9240 - recall: 0.5979 - val_loss: 0.7338 - val_auc: 0.9781 - val_acc: 0.8831 - val_recall: 0.6913
Epoch 15/30
15/15 [==============================] - 3s 186ms/step - loss: 0.7030 - auc: 0.9926 - acc: 0.9272 - recall: 0.6501 - val_loss: 0.7120 - val_auc: 0.9785 - val_acc: 0.8745 - val_recall: 0.7036
Epoch 16/30
15/15 [==============================] - 3s 172ms/step - loss: 0.6612 - auc: 0.9942 - acc: 0.9383 - recall: 0.7092 - val_loss: 0.6682 - val_auc: 0.9794 - val_acc: 0.8844 - val_recall: 0.7651
Epoch 17/30
15/15 [==============================] - 3s 178ms/step - loss: 0.6342 - auc: 0.9942 - acc: 0.9361 - recall: 0.7261 - val_loss: 0.6506 - val_auc: 0.9800 - val_acc: 0.8831 - val_recall: 0.7958
```

Output without `workers`:
```
Epoch 1/30
15/15 [==============================] - 2s 147ms/step - loss: 1.8630 - auc: 0.6431 - acc: 0.2934 - recall: 0.0000e+00 - val_loss: 1.7109 - val_auc: 0.8120 - val_acc: 0.4157 - val_recall: 0.0000e+00
Epoch 2/30
15/15 [==============================] - 2s 121ms/step - loss: 1.7035 - auc: 0.8092 - acc: 0.4322 - recall: 0.0000e+00 - val_loss: 1.5561 - val_auc: 0.8958 - val_acc: 0.5990 - val_recall: 0.0000e+00
Epoch 3/30
15/15 [==============================] - 2s 119ms/step - loss: 1.5716 - auc: 0.8886 - acc: 0.5889 - recall: 0.0000e+00 - val_loss: 1.4155 - val_auc: 0.9392 - val_acc: 0.7700 - val_recall: 0.0000e+00
Epoch 4/30
15/15 [==============================] - 2s 122ms/step - loss: 1.4564 - auc: 0.9276 - acc: 0.6908 - recall: 0.0000e+00 - val_loss: 1.3064 - val_auc: 0.9517 - val_acc: 0.7970 - val_recall: 0.0000e+00
Epoch 5/30
15/15 [==============================] - 2s 121ms/step - loss: 1.3430 - auc: 0.9510 - acc: 0.7625 - recall: 0.0000e+00 - val_loss: 1.2184 - val_auc: 0.9596 - val_acc: 0.8106 - val_recall: 0.0000e+00
Epoch 6/30
15/15 [==============================] - 2s 117ms/step - loss: 1.2485 - auc: 0.9634 - acc: 0.8106 - recall: 0.0000e+00 - val_loss: 1.1361 - val_auc: 0.9648 - val_acc: 0.8364 - val_recall: 0.0000e+00
Epoch 7/30
15/15 [==============================] - 2s 117ms/step - loss: 1.1662 - auc: 0.9717 - acc: 0.8449 - recall: 0.0000e+00 - val_loss: 1.0685 - val_auc: 0.9677 - val_acc: 0.8549 - val_recall: 0.0406
Epoch 8/30
15/15 [==============================] - 2s 120ms/step - loss: 1.0923 - auc: 0.9776 - acc: 0.8628 - recall: 0.0332 - val_loss: 1.0017 - val_auc: 0.9722 - val_acc: 0.8696 - val_recall: 0.2989
Epoch 9/30
15/15 [==============================] - 2s 120ms/step - loss: 1.0200 - auc: 0.9828 - acc: 0.8860 - recall: 0.1778 - val_loss: 0.9470 - val_auc: 0.9729 - val_acc: 0.8758 - val_recall: 0.3530
Epoch 10/30
15/15 [==============================] - 2s 120ms/step - loss: 0.9594 - auc: 0.9845 - acc: 0.8976 - recall: 0.2755 - val_loss: 0.9071 - val_auc: 0.9718 - val_acc: 0.8721 - val_recall: 0.3838
Epoch 11/30
15/15 [==============================] - 2s 118ms/step - loss: 0.9050 - auc: 0.9866 - acc: 0.9145 - recall: 0.3383 - val_loss: 0.8532 - val_auc: 0.9757 - val_acc: 0.8831 - val_recall: 0.4330
Epoch 12/30
15/15 [==============================] - 2s 120ms/step - loss: 0.8490 - auc: 0.9884 - acc: 0.9182 - recall: 0.3947 - val_loss: 0.8108 - val_auc: 0.9777 - val_acc: 0.8844 - val_recall: 0.4859
Epoch 13/30
15/15 [==============================] - 2s 118ms/step - loss: 0.8098 - auc: 0.9891 - acc: 0.9256 - recall: 0.4464 - val_loss: 0.7805 - val_auc: 0.9761 - val_acc: 0.8979 - val_recall: 0.5424
Epoch 14/30
15/15 [==============================] - 2s 123ms/step - loss: 0.7637 - auc: 0.9905 - acc: 0.9261 - recall: 0.5124 - val_loss: 0.7494 - val_auc: 0.9766 - val_acc: 0.8844 - val_recall: 0.6273
Epoch 15/30
15/15 [==============================] - 2s 122ms/step - loss: 0.7184 - auc: 0.9925 - acc: 0.9356 - recall: 0.5704 - val_loss: 0.7221 - val_auc: 0.9767 - val_acc: 0.8893 - val_recall: 0.7023
Epoch 16/30
15/15 [==============================] - 2s 124ms/step - loss: 0.6860 - auc: 0.9929 - acc: 0.9372 - recall: 0.6232 - val_loss: 0.6965 - val_auc: 0.9766 - val_acc: 0.8844 - val_recall: 0.7343
Epoch 17/30
15/15 [==============================] - 2s 123ms/step - loss: 0.6484 - auc: 0.9926 - acc: 0.9325 - recall: 0.6960 - val_loss: 0.6632 - val_auc: 0.9785 - val_acc: 0.8918 - val_recall: 0.7724
Epoch 18/30
15/15 [==============================] - 2s 114ms/step - loss: 0.6211 - auc: 0.9944 - acc: 0.9372 - recall: 0.7298 - val_loss: 0.6538 - val_auc: 0.9772 - val_acc: 0.8844 - val_recall: 0.7712
```

</details>
